go/store/datas/puller.go

```diff
@@ -0,0 +1,294 @@
+	if srcDB.chunkStore().Version() != sinkDB.chunkStore().Version() {
+		return nil, fmt.Errorf("cannot pull from src to sink; src version is %v and sink version is %v", srcDB.chunkStore().Version(), sinkDB.chunkStore().Version())
+	}
+
+	wr, err := nbs.NewCmpChunkTableWriter()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Puller{
+		fmt:           srcDB.Format(),
+		srcDB:         srcDB,
+		sinkDB:        sinkDB,
+		rootChunkHash: rootChunkHash,
+		downloaded:    hash.HashSet{},
+		tempDir:       tempDir,
+		wr:            wr,
+		chunksPerTF:   chunksPerTF,
+	}, nil
+}
+
+func (p *Puller) processCompletedTables(ctx context.Context, ae *atomicerr.AtomicError, completedTables <-chan FilledWriters) {
+	type tempTblFile struct {
+		id        string
+		path      string
+		numChunks int
+	}
+
+	var tblFiles []tempTblFile
+
+	var err error
+	for tblFile := range completedTables {
+		if err != nil {
+			continue // drain
+		}
+
+		var id string
+		id, err = tblFile.wr.Finish()
+
+		if ae.SetIfError(err) {
+			continue
+		}
+
+		path := filepath.Join(p.tempDir, id)
+		err = tblFile.wr.FlushToFile(path)
+
+		if ae.SetIfError(err) {
+			continue
+		}
+
+		tblFiles = append(tblFiles, tempTblFile{id, path, tblFile.wr.Size()})
+	}
+
+	for i := len(tblFiles) - 1; i >= 0; i-- {
+		tmpTblFile := tblFiles[i]
+
+		f, err := os.Open(tmpTblFile.path)
+
+		if ae.SetIfError(err) {
+			return
+		}
+
+		err = p.sinkDB.chunkStore().(nbs.TableFileStore).WriteTableFile(ctx, tmpTblFile.id, tmpTblFile.numChunks, f)
+
+		go func() {
+			_ = os.Remove(tmpTblFile.path)
+		}()
+
+		if ae.SetIfError(err) {
+			return
+		}
+	}
+}
+
+// Pull executes the sync operation
+func (p *Puller) Pull(ctx context.Context) error {
+	leaves := make(hash.HashSet)
+	absent := make(hash.HashSet)
+	absent.Insert(p.rootChunkHash)
+
+	ae := atomicerr.New()
+	wg := &sync.WaitGroup{}
+	completedTables := make(chan FilledWriters, 8)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		p.processCompletedTables(ctx, ae, completedTables)
+	}()
+
+	for len(absent) > 0 {
+		var err error
+		absent, err = p.sinkDB.chunkStore().HasMany(ctx, absent)
+
+		if len(absent) > 0 {
+			leaves, absent, err = p.getCmp(ctx, leaves, absent, completedTables)
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if p.wr.Size() > 0 {
+		completedTables <- FilledWriters{p.wr}
+	}
+
+	close(completedTables)
+
+	wg.Wait()
+	return ae.Get()
+}
+
+func (p *Puller) getCmp(ctx context.Context, leaves, batch hash.HashSet, completedTables chan FilledWriters) (hash.HashSet, hash.HashSet, error) {
+	found := make(chan chunks.Chunkable, 4096)
+	processed := make(chan CmpChnkAndRefs, 4096)
+
+	ae := atomicerr.New()
+	go func() {
+		defer close(found)
+		err := p.srcDB.chunkStore().GetManyCompressed(ctx, batch, found)
+		ae.SetIfError(err)
+	}()
+
+	batchSize := len(batch)
+	numChunkWorkers := (batchSize / 1024) + 1
+	if numChunkWorkers > maxChunkWorkers {
+		numChunkWorkers = maxChunkWorkers
+	}
+
+	go func() {
+		defer func() {
```

Comment:
Can't you just defer close(processed)?

Reply:
yes.  I had some debug code in here at one point.  Should have killed the func when I killed that code.  fixed.